File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/LinkMutatorFactory.java
package com.veyndan.paper.reddit.post.media.mutator;

import com.veyndan.paper.reddit.api.reddit.model.PostHint;
import com.veyndan.paper.reddit.post.media.model.Link;
import com.veyndan.paper.reddit.post.model.Post;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

final class LinkMutatorFactory implements MutatorFactory {

    static LinkMutatorFactory create() {
        return new LinkMutatorFactory();
    }

    private LinkMutatorFactory() {
    }

    @Override
    public Maybe<Post> mutate(final Post post) {
        return Single.just(post)
                .filter(post1 -> post1.postHint().value != PostHint.SELF)
                .map(post1 -> {
                    final Link link = Link.create(post1.domain());
                    return post1.withMedias(post1.medias().value.concatWith(Observable.just(link)));
                });
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/LinkMutatorFactory.kt
package com.veyndan.paper.reddit.post.media.mutator

import com.veyndan.paper.reddit.api.reddit.model.PostHint
import com.veyndan.paper.reddit.post.media.model.Link
import com.veyndan.paper.reddit.post.model.Post
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

class LinkMutatorFactory : MutatorFactory {

    override fun mutate(post: Post): Maybe<Post> {
        return Single.just(post)
                .filter { it.postHint != PostHint.SELF }
                .map {
                    val link: Link = Link(it.domain!!)
                    it.copy(it.medias.concatWith(Observable.just(link)))
                }
    }
}
